### PR TITLE
Alex post - Property based testing

### PR DESCRIPTION
**Pull request by 11PJ11**
Property based testing with FsCheck, Resharper and NCrunch in Visual Studio 2013

### CODE REVIEW

File: _posts/2015-12-01-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+Let me know what is your experience with property based testing in F# writing at alex@codurance.com.
+
+### Resources:
+
+- FsCheck [FsCheck]
+- NCrunch [NCrunch]
+- QuickCheck [QuickCheck]
+- Testing with F# Chap. 11 [Testing-FSharp]
+
+[FSCheck]: <http://fscheck.github.io/FsCheck/index.html>
```
Comment (neomatrix369) [inline _posts/2015-12-01-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md]: Merge this below with the above and use markdown for hyperlinks, otherwise the formatting might not end up right.

File: _posts/2015-12-01-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    ``` fs
+            type ListProperties =  
+                static member reverseOfReverseIsAsTheOriginal (xs:int[]) = 
+                    Array.rev (Array.rev xs) = xs
+    ```
+17. Re-run the tests that should be all green by now and read the output. You'll notice it's very verbose as expected and shows all the inputs used to verify the property.
+
+### Conclusion
+You are now ready to experiment with FsCheck!
+Let me know what is your experience with property based testing in F# writing at alex@codurance.com.
```
Comment (neomatrix369) [inline _posts/2015-12-01-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md]: instead of an email, how about twitter handle - everyone can see and get involved in the conversation ;)

File: _posts/2015-12-01-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+---
+
+# Property based testing with FsCheck, Resharper and NCrunch in Visual Studio 2013
+
+### Background
+Craftsmanship in Software strives for high, predictable, repeatable and affordable quality; this is the reason why we care about design, code reviews and of course testing. 
+Test Driven Design helps us create a flexible implementation that meets the expectations of our clients in terms of features, performance and quality; as professionals we have to master as many tools as possible to reach our goal of steadily providing value. 
+	
+### Property based testing
+The above is the reason why I am exploring Property Based Testing; I like to think about it as *"Let the machine do the heavy lifting for you"* or as [John-Huges] says *"Don't write tests, generate them"*.
```
Comment (sandromancuso) [inline _posts/2015-12-01-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md]: Missing link for John-Huges

File: _posts/2015-12-01-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+3. Comment out the content in **FsCheckAddin.fs** file to allow NCrunch to run the tests
+    ``` fs
+        //[<NUnitAddin(Description = "FsCheck addin")>] 
+		//type FsCheckAddin() = 
+		//  interface IAddin with 
+		//      override x.Install host = 
+		//          let tcBuilder = new FsCheckTestCaseBuilder() 
+		//          host.GetExtensionPoint("TestCaseBuilders").Install(tcBuilder) 
+        //          true 
+   ```
```
Comment (sandromancuso) [inline _posts/2015-12-01-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md]: Content above showing on a single line. Try removing space between ``` and fs

Comment (11PJ11) [inline _posts/2015-12-01-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md]: The problem was that I needed a blank line before the ```
Try refreshing plz
I am still working on the links

2015-11-25 17:45 GMT+00:00 Sandro Mancuso notifications@github.com:

> In
> _posts/2015-12-01-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md
> https://github.com/codurance/site/pull/78#discussion_r45897308:
> 
> > +
> > +### Steps
> > +1. Open Visual Studio and create an F# Project
> > +2. Open the Nuget Package Manager following _TOOLS > Nuget Package Manager > Package Manager Console_ from the main menu and type
> > -    1. `Install-Package FsCheck`
> > -    2. `Install-Package FsCheck.Nunit`
> >   +3. Comment out the content in **FsCheckAddin.fs** file to allow NCrunch to run the tests
> > -    ``` fs
> > -        //[<NUnitAddin(Description = "FsCheck addin")>]
> > -       //type FsCheckAddin() =
> > -       //  interface IAddin with
> > -       //      override x.Install host =
> > -       //          let tcBuilder = new FsCheckTestCaseBuilder()
> > -       //          host.GetExtensionPoint("TestCaseBuilders").Install(tcBuilder)
> > -        //          true
> > -   ```
> 
> Content above showing on a single line. Try removing space between ``` and
> fs
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/codurance/site/pull/78/files#r45897308.

## 

alessandro di gioia

---

Il presente messaggio di posta elettronica è ad esclusivo utilizzo del
destinatario indicato in indirizzo e gli eventuali documenti allegati
potrebbero avere carattere riservato. Qualora non foste il destinatario del
presente messaggio Vi preghiamo di non leggerlo, di cancellarlo dal Vostro
sistema assieme ad ogni documento ad esso allegato e di volerci avvertire
immediatamente tramite posta elettronica o telefonicamente.

This e-mail is for the sole use of the intended recipient and any file
transmitted with it may contain material that is confidential and
privileged. If you are not the intended recipient of this e-mail, please do
not read this e-mail and delete this message and any file attached from
your system and then notify us immediately by reply e-mail or by telephone.

**Comment by dhatanian** [comment]
Great article ! Can you do another one in ruby and one in nodejs ? :smiling_imp:

**Comment by 11PJ11** [comment]
I need to write a blog post on how to install them on Windows first

2015-11-25 18:16 GMT+00:00 David Hatanian notifications@github.com:

> Great article ! Can you do another one in ruby and one in nodejs ? [image:
> :smiling_imp:]
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/codurance/site/pull/78#issuecomment-159691980.

## 

alessandro di gioia

---

Il presente messaggio di posta elettronica è ad esclusivo utilizzo del
destinatario indicato in indirizzo e gli eventuali documenti allegati
potrebbero avere carattere riservato. Qualora non foste il destinatario del
presente messaggio Vi preghiamo di non leggerlo, di cancellarlo dal Vostro
sistema assieme ad ogni documento ad esso allegato e di volerci avvertire
immediatamente tramite posta elettronica o telefonicamente.

This e-mail is for the sole use of the intended recipient and any file
transmitted with it may contain material that is confidential and
privileged. If you are not the intended recipient of this e-mail, please do
not read this e-mail and delete this message and any file attached from
your system and then notify us immediately by reply e-mail or by telephone.

File: _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        --- Checking ListProperties ---
+		System.Reflection.TargetInvocationException : Exception has been thrown by the target of an invocation.
+		  ----> System.Exception : ListProperties.ReverseOfReverseIsAsTheOriginal-Falsifiable, after 6 tests (3 shrinks) (StdGen (1165808905,296086422)):
+		Original:
+		[-1; 2; -2]
+		Shrunk:
+		[1; 0]
+    ```
+	
+12. This shows that the property was falsifiable (as expected) but let's try to gather more informations to address the issue
```
Comment (neomatrix369) [inline _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md:114]: information and not information**s**

File: _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+	<center><img src="/assets/img/custom/blog/2015-11-25-Property-based-testing/Resharper-tests-run-error.png"></center>
+11. You should read a message similar to
+    
+	```
+        --- Checking ListProperties ---
+		System.Reflection.TargetInvocationException : Exception has been thrown by the target of an invocation.
+		  ----> System.Exception : ListProperties.ReverseOfReverseIsAsTheOriginal-Falsifiable, after 6 tests (3 shrinks) (StdGen (1165808905,296086422)):
+		Original:
+		[-1; 2; -2]
+		Shrunk:
```
Comment (neomatrix369) [inline _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md:110]: it would be nice to explain what the term Shrunk means.

File: _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+		shrink:
+		[6; 0]
+		shrink:
+		[3; 0]
+		shrink:
+		[2; 0]
+		shrink:
+		[1; 0]
+		System.Reflection.TargetInvocationException : Exception has been thrown by the target of an invocation.
+		  ----> System.Exception : ListProperties.ReverseOfReverseIsAsTheOriginal-Falsifiable, after 6 tests (11 shrinks) (StdGen (1681135586,296086423)):
```
Comment (neomatrix369) [inline _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md:162]: however trivial it might sound, please explain 'Falsifiable' or put a link to it.

File: _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+15. The output shows the "shrinking process" from the sixth input (identified by 5:) original value ```[-3; 1; 5; -3; 6; 2; 6; -2]``` to the  "shrunk" input ```[1; 0]``` that still falsifies the property 
+16. Let's correct the property as **the reverse of the reverse of a list is equal to the original**
+    
+	``` fs
+            type ListProperties =
+				static member ReverseOfReverseIsAsTheOriginal (xs:int list) = 
+					List.rev (List.rev xs) = xs
+    ```
+	
+17. Re-run the tests that should be all green by now and read the output. You'll notice it's very verbose as expected and shows all the inputs used to verify the property.
```
Comment (neomatrix369) [inline _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md:178]: Show the screen-shot with successful execution.

**Comment by neomatrix369** [comment]
Program or component names like 'FsCheck' etc... across the post could be highlighted in italics or bold or another colouration.

**Comment by neomatrix369** [comment]
:+1: very good flow and helps the reader get to the point with installing the components required to do property-base testing.

**Comment by 11PJ11** [comment]
Hi all is this post good enough for production?

**Comment by dhatanian** [comment]
Yes !

**Comment by neomatrix369** [comment]
Looks good for me - someone please press the `merge` button ;)

File: _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+- property based testing
+- visual studio
+---
+
+### Background
+Craftsmanship in software strives for high, predictable, repeatable and affordable quality; this is the reason why we care about design, code reviews and of course testing. 
+Test Driven Development helps us create a flexible implementation that meets the expectations of our clients in terms of features, performance and quality; as professionals we have to master as many tools as possible to reach our goal of steadily providing value. 
+	
+### Property based testing
+The above is the reason why I am exploring Property Based Testing; I like to think about it as *"Let the machine do the heavy lifting for you"* or as [John Huges] says *"Don't write tests, generate them"*.
```
Comment (TobyRet) [inline _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md]: Incorrect use of semi colons throughout (!). Here is a guide on when to use. https://writing.wisc.edu/Handbook/Semicolons.html

File: _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+I am not claiming that you should stop writing unit tests, I find that when driving the design [Outside-in] hence declaring the relations between my components, unit tests are the tool to use.
```
Comment (TobyRet) [inline _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md]: I am not claiming that you should stop writing unit tests - [use an em-dash]
I find that when driving the design 'outside-in' and declaring the relationships between my components, that unit testing is an appropriate tool to use.

File: _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+However I sometimes want to discover deeper properties of my domain, and sometimes I am faced with features that have more edge cases than my brain could handle.
```
Comment (TobyRet) [inline _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md]: Sometimes when I want to discover deeper properties of my domain, I am faced with features that have more edge cases than my brain could handle.

**Comment by tekavec** [comment]
.Net => .NET (as Microsoft uses it: http://www.microsoft.com/net)

File: _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+### Background
+Craftsmanship in software strives for high, predictable, repeatable and affordable quality; this is the reason why we care about design, code reviews and of course testing. 
+Test Driven Development helps us create a flexible implementation that meets the expectations of our clients in terms of features, performance and quality; as professionals we have to master as many tools as possible to reach our goal of steadily providing value. 
+	
+### Property based testing
+The above is the reason why I am exploring Property Based Testing; I like to think about it as *"Let the machine do the heavy lifting for you"* or as [John Huges] says *"Don't write tests, generate them"*.
+
+I am not claiming that you should stop writing unit tests, I find that when driving the design [Outside-in] hence declaring the relations between my components, unit tests are the tool to use.
+However I sometimes want to discover deeper properties of my domain, and sometimes I am faced with features that have more edge cases than my brain could handle. 
+In such cases, generating a wide array of inputs to test my program is appealing; this is where I think property testing becomes a viable tool.
```
Comment (TobyRet) [inline _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md]: Check if this is correct use of the semi-colon

File: _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+	
+5. Add this test to enable [NCrunch]
+    
+	``` fs
+            //Needed to enable NCrunch 
+            [<Ignore>][<Test>] 
+            let ignoreMe() = () 
+    ```
+	
+6. Create a static class whose methods are the properties you want to test. For this example I'll test a plainly wrong property about lists which is **the reverse of a list is always equal to the original**.
```
Comment (TobyRet) [inline _posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md]: For this example I'll test an incorrect implementation of a list - **the reverse of a list is always equal to the original**.
